### PR TITLE
arch: inline up_interrupt_context() 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arch: inline up_interrupt_context()

inline the up_interrupt_context() to avoid unnecessary stack pushes

Signed-off-by: chao.an <anchao@xiaomi.com>

## Impact

N/A

## Testing

ci check

### CODE REVIEW

**Comment by davids5** [comment]
What are the code size and C89 implication?

**Comment by anchao** [comment]
> What are the code size and C89 implication?

@davids5 ,

There is a slight increase in code size, but overall the impact is not very serious
lm3s6965-ek:qemu-flat(cortex-m):

Before:
```
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 222304	    324	  14848	 237476	  39fa4	nuttx
```
 
 After:
```
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 222320	    324	  14848	 237492	  39fb4	nuttx
```


sabre-6quad:smp(Cortex-a9,smp):

Before:
```
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 127840	    196	  33500	 161536	  27700	nuttx
```

After:
```
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 128096	    196	  33500	 161792	  27800	nuttx
```
 
about C89 compatible, It seems that the current code base can no longer be built with c89 normally(without this PR), too many inline keywords, the inline keyword was added in C99, I think we can use the -std=gnu89 flag instead of -std=c89

File: include/nuttx/arch.h [hunk 1]
```diff
@@ -1939,28 +1928,6 @@ int16_t up_fetchsub16(FAR volatile int16_t *addr, int16_t value);
...
- * Input Parameters:
- *   None
- *
- * Returned Value:
- *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
- *   corresponds to the currently executing CPU.
- *
- ****************************************************************************/
-
-#ifdef CONFIG_SMP
```
Comment (xiaoxiang781216) [inline include/nuttx/arch.h:1958]: let's keep the prototype here, you can implement up_cpu_index as inline or not inline function freely with the prototype

Comment (anchao) [reply include/nuttx/arch.h:1958]: The implementation of up_interrupt_context() is different on different arch platform, especially some architectures that do not support SMP yet

File: arch/arm/include/irq.h [hunk 2]
```diff
@@ -29,6 +29,12 @@
  * Included Files
  ****************************************************************************/
 
+#include <nuttx/config.h>
+#ifndef __ASSEMBLY__
+#  include <stdbool.h>
+#  include <sys/types.h>
```
Comment (pkarashchenko) [inline arch/arm/include/irq.h:35]: I think `sys/types.h` can be moved out of `ifndef __ASSEMBLY__`

File: arch/avr/include/irq.h [hunk 3]
```diff
@@ -55,19 +60,25 @@
...
  * Public Data
  ****************************************************************************/
 
+#ifndef __ASSEMBLY__
+/* This holds a references to the current interrupt level register storage
+ * structure.  If is non-NULL only during interrupt processing.
+ */
+
+#ifdef CONFIG_ARCH_FAMILY_AVR32
+extern volatile uint32_t *g_current_regs;
```
Comment (pkarashchenko) [inline arch/avr/include/irq.h:73]:
```suggestion
EXTERN volatile uint32_t *g_current_regs;
```

File: arch/avr/include/irq.h [hunk 4]
```diff
@@ -29,6 +29,11 @@
  * Included Files
  ****************************************************************************/
 
+#ifndef __ASSEMBLY__
+#  include <stdbool.h>
+#  include <sys/types.h>
```
Comment (pkarashchenko) [inline arch/avr/include/irq.h:34]: can move this out of `__ASSEMBLY__`

File: arch/hc/include/irq.h [hunk 5]
```diff
@@ -55,14 +60,38 @@
...
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+#ifndef __ASSEMBLY__
+/* This holds a references to the current interrupt level register storage
+ * structure.  If is non-NULL only during interrupt processing.
+ */
+
+extern volatile uint8_t *g_current_regs;
```
Comment (pkarashchenko) [inline arch/hc/include/irq.h:72]:
```suggestion
EXTERN volatile uint8_t *g_current_regs;
```

File: arch/mips/include/irq.h [hunk 6]
```diff
@@ -55,14 +60,23 @@
...
+ * register storage structure.  It is non-NULL only during interrupt
+ * processing.  Access to g_current_regs must be through the macro
+ * CURRENT_REGS for portability.
+ */
+
+/* For the case of architectures with multiple CPUs, then there must be one
+ * such value for each processor that can receive an interrupt.
+ */
+
+extern volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];
```
Comment (pkarashchenko) [inline arch/mips/include/irq.h:77]:
```suggestion
EXTERN volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];
```

File: arch/arm/include/irq.h (hunk 2 above)
Comment (anchao) [reply arch/arm/include/irq.h]: Done

File: arch/avr/include/irq.h (hunk 3 above)
Comment (anchao) [reply arch/avr/include/irq.h]: Done

File: arch/avr/include/irq.h (hunk 4 above)
Comment (anchao) [reply arch/avr/include/irq.h]: Done

File: arch/arm/include/irq.h [hunk 7]
```diff
@@ -29,6 +29,12 @@
  * Included Files
  ****************************************************************************/
 
+#include <nuttx/config.h>
+#ifndef __ASSEMBLY__
+#  include <stdbool.h>
+#endif
+#include <sys/types.h>
```
Comment (pkarashchenko) [inline arch/arm/include/irq.h:36]: I would better reorganized this in
```suggestion
#include <sys/types.h>
#ifndef __ASSEMBLY__
#  include <stdbool.h>
#endif
```
here and in other places

File: arch/z80/include/irq.h [hunk 8]
```diff
@@ -32,4 +32,55 @@
...
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
+#else
+#  define up_cpu_index() (0)
+#endif
+#endif
```
Comment (pkarashchenko) [inline arch/z80/include/irq.h:84]: Why do we need this in platform files if we have it already present in `include/nuttx/arch.h`?

Comment (anchao) [reply arch/z80/include/irq.h:84]: irq.h will be included before arch.h during preprocessing phase

```
In file included from incubator-nuttx/include/nuttx/irq.h:37,
                 from incubator-nuttx/include/nuttx/sched.h:39,
                 from incubator-nuttx/include/nuttx/arch.h:87,
                 from incubator-nuttx/include/nuttx/userspace.h:35,
                 from n2/incubator-nuttx/include/nuttx/kmalloc.h:36,
                 from incubator-nuttx/include/nuttx/lib/lib.h:31,
                 from libxx_cxa_atexit.cxx:27:
incubator-nuttx/include/arch/irq.h: In function 'bool up_interrupt_context()':
incubator-nuttx/include/arch/irq.h:135:14: error: 'up_cpu_index' was not declared in this scope
  135 |   bool ret = CURRENT_REGS != NULL;
      |              ^~~~~~~~~~~~
```

File: arch/z80/include/irq.h [hunk 9]
```diff
@@ -32,4 +32,55 @@
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/z80/include/irq.h:80]: don't need this section if arch doesn't support SMP.

File: arch/z16/include/irq.h [hunk 10]
```diff
@@ -57,6 +62,61 @@ extern "C"
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/z16/include/irq.h:115]: remove

File: arch/x86_64/include/irq.h [hunk 11]
```diff
@@ -74,6 +78,48 @@ extern "C"
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/x86_64/include/irq.h:98]: remove

File: arch/x86/include/irq.h [hunk 12]
```diff
@@ -54,24 +58,70 @@
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/x86/include/irq.h:102]: remove

File: arch/sparc/include/irq.h [hunk 13]
```diff
@@ -55,24 +60,66 @@
...
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
```
Comment (xiaoxiang781216) [inline arch/sparc/include/irq.h:119]: remove

File: arch/sim/include/irq.h [hunk 14]
```diff
@@ -77,6 +125,7 @@ extern "C"
 
 irqstate_t up_irq_save(void);
 void up_irq_restore(irqstate_t flags);
+bool up_interrupt_context(void);
```
Comment (xiaoxiang781216) [inline arch/sim/include/irq.h:128]: revert the change if up_interrutp_context isn't linlined

File: arch/risc-v/include/irq.h [hunk 15]
```diff
@@ -612,18 +661,32 @@ static inline void up_irq_restore(irqstate_t flags)
...
- *   Return the current interrupt state and enable interrupts
+ *   Return true is we are currently executing in the interrupt
+ *   handler context.
  *
  ****************************************************************************/
 
-EXTERN irqstate_t up_irq_enable(void);
+static inline bool up_interrupt_context(void)
+{
+#ifdef CONFIG_ARCH_RV64
```
Comment (xiaoxiang781216) [inline arch/risc-v/include/irq.h:674]: why check rv54?

File: arch/renesas/include/irq.h [hunk 16]
```diff
@@ -40,27 +45,70 @@
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/renesas/include/irq.h:89]: remove

File: arch/misoc/include/irq.h [hunk 17]
```diff
@@ -56,6 +57,60 @@ extern "C"
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/misoc/include/irq.h:109]: revert the change

File: arch/hc/include/irq.h [hunk 18]
```diff
@@ -55,25 +60,67 @@
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/hc/include/irq.h:121]: revert

File: arch/ceva/include/irq.h [hunk 19]
```diff
@@ -75,4 +80,80 @@
...
+ *
+ * Description:
+ *   Return true is we are currently executing in the interrupt
+ *   handler context.
+ *
+ ****************************************************************************/
+
+static inline bool up_interrupt_context(void)
+{
+  return CURRENT_REGS != NULL;
```
Comment (xiaoxiang781216) [inline arch/ceva/include/irq.h:150]: need irqsave here like arm

File: arch/avr/include/irq.h [hunk 20]
```diff
@@ -55,27 +60,73 @@
...
+ *   None
+ *
+ * Returned Value:
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/avr/include/irq.h:107]: revert

Comment (anchao) [reply arch/avr/include/irq.h]: Done

File: arch/ceva/include/irq.h (hunk 19 above)
Comment (anchao) [reply arch/ceva/include/irq.h]: Done

File: arch/hc/include/irq.h (hunk 18 above)
Comment (anchao) [reply arch/hc/include/irq.h]: Done

File: arch/z80/include/irq.h (hunk 8 above)
Comment (pkarashchenko) [reply arch/z80/include/irq.h]: can we include `nuttx/arch.h` right after `#include <nuttx/irq.h>` in arch irq.h files?

Comment (anchao) [reply arch/z80/include/irq.h]: In this case, a lot of include logic needs to be changed. up_cpu_index() is just a declaration in arch.h. If there is no special requirement for up_cpu_index() in the platform code, this function can be defined only in arch.h without repetition from irq.h

File: arch/x86/include/irq.h (hunk 12 above)
Comment (xiaoxiang781216) [reply arch/x86/include/irq.h:102]: But, no change?

File: arch/z80/include/irq.h [hunk 21]
```diff
@@ -32,4 +32,33 @@
 #include <nuttx/irq.h>
 #include <arch/chip/irq.h>
 
+#ifndef __ASSEMBLY__
+#  include <stdbool.h>
+#  include "chip/switch.h"
```
Comment (xiaoxiang781216) [inline arch/z80/include/irq.h:37]: it isn't good to include the arch specific chip file in the common header file.

File: arch/sim/include/irq.h [hunk 22]
```diff
@@ -50,20 +50,39 @@ struct xcptcontext
...
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
 
-#ifdef __cplusplus
-#define EXTERN extern "C"
-extern "C"
-{
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
```
Comment (xiaoxiang781216) [inline arch/sim/include/irq.h:83]: where is up_interrupt_context? should we revert the change?

Comment (anchao) [reply arch/sim/include/irq.h:83]: Done

File: arch/or1k/include/irq.h [hunk 23]
```diff
@@ -37,4 +42,80 @@
...
+ * Name: up_interrupt_context
+ *
+ * Description:
+ *   Return true is we are currently executing in the interrupt
+ *   handler context.
+ *
+ ****************************************************************************/
+
+static inline bool up_interrupt_context(void)
+{
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/or1k/include/irq.h:111]: Done

File: arch/x86/include/irq.h (hunk 12 above)
Comment (anchao) [reply arch/x86/include/irq.h:102]: Done

File: arch/arm/include/irq.h [hunk 24]
```diff
@@ -55,4 +60,91 @@
...
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SMP
+int up_cpu_index(void);
+#else
+#  define up_cpu_index() (0)
+#endif
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/arm/include/irq.h:114]: There is no special difference between defining in arch/arch.h or arch/irq.h,
how about delete the definition of nuttx/arch.h and let each arch/irq.h implement it by itself? which will avoids duplicate declarations

[the remark this answers is not in the document]
Comment (pkarashchenko) [reply arch/arm/include/irq.h:114]: IMO `arch/irq.h` is not the proper header as the CPU index is not related to IRQ. I think that `arch/arch.h` is more suitable

[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm/include/irq.h:114]: But it's also strange to include arch/arch.h from arch/irq.h. up_cpu_index is most used in the context switch code, and all related arch staff is defined in arch/irq.h(e.g. xcptcontext).

[the remark this answers is not in the document]
Comment (pkarashchenko) [reply arch/arm/include/irq.h:114]: Ok. Let's keep it in irq.h

[the remark this answers is not in the document]
Comment (anchao) [reply arch/arm/include/irq.h:114]: Done

File: arch/z16/include/irq.h [hunk 25]
```diff
@@ -57,6 +62,61 @@ extern "C"
...
+ *   handler context.
+ *
+ ****************************************************************************/
+
+#ifndef __ASSEMBLY__
+static inline bool up_interrupt_context(void)
+{
+  return g_current_regs != NULL;
+}
+#endif
```
Comment (pkarashchenko) [inline arch/z16/include/irq.h:118]: Let's switch to defines in all possible places
```suggestion
#define up_interrupt_context() (g_current_regs != NULL)
```

Comment (anchao) [reply arch/z16/include/irq.h]: Done

File: arch/mips/include/irq.h [hunk 26]
```diff
@@ -54,27 +59,72 @@
...
+ *   An integer index in the range of 0 through (CONFIG_SMP_NCPUS-1) that
+ *   corresponds to the currently executing CPU.
+ *
  ****************************************************************************/
 
-#ifdef __cplusplus
-#define EXTERN extern "C"
-extern "C"
-{
+#ifdef CONFIG_SMP
```
Comment (pkarashchenko) [inline arch/mips/include/irq.h:107]: Do we have MIPS SMP case?

Comment (anchao) [reply arch/mips/include/irq.h]: Done

File: arch/mips/include/irq.h [hunk 27]
```diff
@@ -54,26 +59,67 @@
...
+ * processing.  Access to g_current_regs must be through the macro
+ * CURRENT_REGS for portability.
+ */
+
+/* For the case of architectures with multiple CPUs, then there must be one
+ * such value for each processor that can receive an interrupt.
+ */
+
+EXTERN volatile uint32_t *g_current_regs[CONFIG_SMP_NCPUS];
+#define CURRENT_REGS (g_current_regs[up_cpu_index()])
```
Comment (pkarashchenko) [inline arch/mips/include/irq.h:85]:
```suggestion
EXTERN volatile uint32_t *g_current_regs;
```

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/mips/include/irq.h (hunk 27 above)
Comment (anchao) [reply arch/mips/include/irq.h:85]: CURRENT_REGS is used in the mips

File: arch/mips/include/irq.h [hunk 28]
```diff
@@ -54,26 +59,67 @@
...
+/****************************************************************************
+ * Name: up_interrupt_context
+ *
+ * Description:
+ *   Return true is we are currently executing in the interrupt
+ *   handler context.
+ *
+ ****************************************************************************/
+
+#define up_interrupt_context() (CURRENT_REGS != NULL)
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/mips/include/irq.h]: Done

**Review by pkarashchenko — APPROVED** [review]
